modules/renter/workerdownload.go

```diff
@@ -1,55 +1,200 @@
 package renter
 
+// workerdownload.go is responsible for coordinating the actual fetching of
+// pieces, determining when to add standby workers, when to perform repairs, and
+// coordinating resource management between the workers operating on a chunk.
+
 import (
-	"github.com/NebulousLabs/Sia/crypto"
-	"github.com/NebulousLabs/Sia/types"
+	"errors"
+	"sync/atomic"
 )
 
-type (
-	// downloadWork contains instructions to download a piece from a host, and
-	// a channel for returning the results.
-	downloadWork struct {
-		// dataRoot is the MerkleRoot of the data being requested, which serves
-		// as an ID when requesting data from the host.
-		dataRoot   crypto.Hash
-		pieceIndex uint64
+// cleanUp will check if the download has failed, and if not it will add any
+// standby workers which need to be added. Calling cleanUp too many times is not
+// harmful, however missing a call to cleanUp can lead to dealocks.
+//
+// NOTE: The fact that calls to cleanUp must be actively managed is probably the
+// weakest part of the design of the download code.
+func (udc *unfinishedDownloadChunk) cleanUp() {
+	// Return any unused memory.
+	udc.returnMemory()
 
-		chunkDownload *chunkDownload
+	// Check if the chunk has failed. If so, fail the download and return any
+	// remaining memory.
+	if udc.workersRemaining+udc.piecesCompleted < udc.erasureCode.MinPieces() && !udc.failed {
+		udc.fail(errors.New("not enough workers to continue download"))
+		return
+	}
 
-		// resultChan is a channel that the worker will use to return the
-		// results of the download.
-		resultChan chan finishedDownload
+	// Check if standby workers are required, and add them if so.
+	chunkComplete := udc.piecesCompleted >= udc.erasureCode.MinPieces()
+	desiredPiecesRegistered := udc.erasureCode.MinPieces() + udc.staticOverdrive - udc.piecesCompleted
+	if !chunkComplete && udc.piecesRegistered < desiredPiecesRegistered {
+		for i := 0; i < len(udc.workersStandby); i++ {
+			udc.workersStandby[i].managedQueueDownloadChunk(udc)
+		}
+		udc.workersStandby = udc.workersStandby[:0]
 	}
+}
 
-	// finishedDownload contains the data and error from performing a download.
-	finishedDownload struct {
-		chunkDownload *chunkDownload
-		data          []byte
-		err           error
-		pieceIndex    uint64
-		workerID      types.FileContractID
+// returnMemory will check on the status of all the workers and pieces, and
+// determine how much memory is safe to return to the renter. This should be
+// called each time a worker returns, and also after the chunk is recovered.
+func (udc *unfinishedDownloadChunk) returnMemory() {
+	// The maximum amount of memory is the pieces completed plus the number of
+	// workers remaining.
+	initialMemory := uint64(udc.staticOverdrive+udc.erasureCode.MinPieces()) * udc.staticPieceSize
+	maxMemory := uint64(udc.workersRemaining+udc.piecesCompleted) * udc.staticPieceSize
```

Comment:
Should we sanity check for `udc.workersRemaining+udc.piecesCompleted < udc.erasureCode.MinPieces()` ?

Reply:
no, because you might still have overdrive pieces. So even if the download has completed, there may still be memory that's tied up and can't be released.